### PR TITLE
Removing the glNaked code for capturing the billboard snaphot and rendering the rearViewMirror in a separate render target

### PR DESCRIPTION
**Pull request by samcake**
- Introduce the gpu::Context::downloadFramebuffer call allowing to capture a system meory version of the framebuffer. very useful for doing a snapshot but not very efficient.

- Separated the rendering of the rear view mirror from the main rendering with the scene, it s now happening in a separate framebuffer (the selfie framebuffer) and displayed as a normal textured quad during the application Overlay pass

- Found a nasty bug with the scissor rect not assigned properly, this solve a bunch of issues

- Fixed a bug when clearing a color buffer of Framebuffer, we force the COlorMAsk to WRITE_ALL

- Introduced a new Job in the rendering pipeline to take care of the framebuffer setup and first clear on depth and color 0. THis simplify the usage of displaySIde / render engine since no need to setup primaryBUffer and do the clear. This shows in renderRearViewMirror and PaintGL mostly

- introduced a primaryFramebufferDepthCOlor only to avoidhaving to use _glDrawBuffers and consider Framebuffers immutable like they should be. 
-- A question would be, should we have a FramebufferView that would expose the active buffers ?

### CODE REVIEW

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -936,7 +963,7 @@ void Application::paintGL() {
     }
 
     // Sync up the View Furstum with the camera
-    loadViewFrustum(_myCamera, _viewFrustum);
+ //   loadViewFrustum(_myCamera, _viewFrustum);
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: dead code

File: interface/src/avatar/MyAvatar.cpp [hunk 2]
```diff
@@ -1551,6 +1551,7 @@ void MyAvatar::maybeUpdateBillboard() {
     QBuffer buffer(&_billboard);
     buffer.open(QIODevice::WriteOnly);
     image.save(&buffer, "PNG");
+    image.save("billboard.png", "PNG");
```
Comment (ZappoMan) [inline interface/src/avatar/MyAvatar.cpp]: dead code??

**Comment by hifi-gustavo** [comment]
Build console output for commit: a002560003f484e4e49f716597331d46f8fa3d46 available at: https://gist.github.com/fee9f2c1417f6731faf7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.dmg

File: interface/src/Application.cpp [hunk 3]
```diff
@@ -875,6 +875,33 @@ void Application::paintGL() {
...
     resizeGL();
 
+    // Before anything else, let's sync up the gpuContext with the true glcontext used in case anything happened
+    renderArgs._context->syncCache();
+
+    if (Menu::getInstance()->isOptionChecked(MenuOption::Mirror)) {
+        auto primaryFbo = DependencyManager::get<FramebufferCache>()->getPrimaryFramebuffer();
+        {
+            gpu::Batch batch;
+            batch.setFramebuffer(nullptr);
```
Comment (jherico) [inline interface/src/Application.cpp]: why the multiple calls to setFramebuffer?

File: interface/src/Application.cpp [hunk 4]
```diff
@@ -2973,25 +2998,31 @@ PickRay Application::computePickRay(float x, float y) const {
...
-                               BILLBOARD_SIZE, BILLBOARD_SIZE),
-                         true);
+    _glWidget->makeCurrent();
+
+    auto primaryFbo = DependencyManager::get<FramebufferCache>()->getPrimaryFramebuffer();
+
+    {
+        gpu::Batch batch;
+        batch.setFramebuffer(nullptr);
+        batch.setFramebuffer(primaryFbo);
```
Comment (jherico) [inline interface/src/Application.cpp]: again with two calls.... is this necessary?

**Comment by ZappoMan** [comment]
test this please

File: libraries/gpu/src/gpu/GLBackendOutput.cpp [hunk 5]
```diff
@@ -11,6 +11,8 @@
 #include "GPULogging.h"
 #include "GLBackendShared.h"
 
+#include "qimage.h"
```
Comment (jherico) [inline libraries/gpu/src/gpu/GLBackendOutput.cpp]: should be <QImage> and come before the other headers.

**Comment by hifi-gustavo** [comment]
Build console output for commit: a002560003f484e4e49f716597331d46f8fa3d46 available at: https://gist.github.com/fee9f2c1417f6731faf7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.dmg

**Comment by ZappoMan** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: a002560003f484e4e49f716597331d46f8fa3d46 available at: https://gist.github.com/fee9f2c1417f6731faf7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.dmg

File: interface/src/Application.cpp (hunk 3 above)
Comment (samcake) [inline interface/src/Application.cpp]: THe first one make sure we are in sync with the cache, this is something i need to handle better , there are cases when the cache is not in sync with the true glFramebuffer bound underneath. By doing that i guarantee it s going tobe set correctly
i was planning to fix that next indeed...

File: libraries/gpu/src/gpu/GLBackendOutput.cpp (hunk 5 above)
Comment (samcake) [inline libraries/gpu/src/gpu/GLBackendOutput.cpp]: ok

**Comment by ZappoMan** [comment]
ok - you definitely have legit build errors...

```
70>d:\jenkins\workspace\pull-request-win\libraries\gpu\src\gpu\Framebuffer.h(135): error C2061: syntax error : identifier 'QImage' (..\..\..\libraries\gpu\src\gpu\Batch.cpp) [D:\jenkins\workspace\pull-request-win\build\libraries\gpu\gpu.vcxproj]
    70>d:\jenkins\workspace\pull-request-win\libraries\gpu\src\gpu\Framebuffer.h(135): error C2061: syntax error : identifier 'QImage' (..\..\..\libraries\gpu\src\gpu\Context.cpp) [D:\jenkins\workspace\pull-request-win\build\libraries\gpu\gpu.vcxproj]
    70>d:\jenkins\workspace\pull-request-win\libraries\gpu\src\gpu\Framebuffer.h(135): error C2061: syntax error : identifier 'QImage' (..\..\..\libraries\gpu\src\gpu\Framebuffer.cpp) [D:\jenkins\workspace\pull-request-win\build\libraries\gpu\gpu.vcxproj]
```

**Comment by ZappoMan** [comment]
and...
```
In file included from /Users/jenkins/Jenkins/workspace/pull-request-osx/libraries/gpu/src/gpu/Batch.cpp:11:
In file included from /Users/jenkins/Jenkins/workspace/pull-request-osx/libraries/gpu/src/gpu/Batch.h:16:
/Users/jenkins/Jenkins/workspace/pull-request-osx/libraries/gpu/src/gpu/Framebuffer.h:135:19: error: unknown type name 'QImage'
    void getImage(QImage* result) const;
                  ^
```

**Comment by hifi-gustavo** [comment]
Build console output for commit: a424ff98ca06157d94ffa04c9be9221bf1f54d08 available at: https://gist.github.com/fee9f2c1417f6731faf7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.dmg

**Comment by samcake** [comment]
ABORT , it s broken on MAC

**Comment by hifi-gustavo** [comment]
Build console output for commit: ded95a3dde482f21d48577d370559561b3f91fbf available at: https://gist.github.com/fee9f2c1417f6731faf7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8421aafe33efc21215c2cc67bbe82f192365a43f available at: https://gist.github.com/fee9f2c1417f6731faf7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 66cd17ec6663e0800e20753e8a552ee0a7e1d94f available at: https://gist.github.com/fee9f2c1417f6731faf7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: cb375e8ecee5ce439e2714eb3635bfa44050c1c4 available at: https://gist.github.com/fee9f2c1417f6731faf7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.dmg

**Comment by samcake** [comment]
On MAC still having an issue if the rearview mirror is visible during resize
The drawing of the selfieImage draw garbage in the overlay map if it s happening right after the resize.
it s working fine on windows...

**Comment by hifi-gustavo** [comment]
Build console output for commit: 52ac5b3ef63331d193f6abe4c8bab39a1410b3d4 available at: https://gist.github.com/fee9f2c1417f6731faf7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.dmg

**Comment by jherico** [comment]
screen shot 2015-07-24 at 11 24 36 am

Overlay is not properly being cleared.  If you open and close something on the overlay it leave a shadow.  Looks like the alpha channel is not being cleared.

**Comment by jherico** [comment]
The problem goes away if I put `glColorMask(true, true, true, true);` at the start of the `ApplicationOverlay::renderOverlay` method.  So something is not properly setting the color mask.

**Comment by samcake** [comment]
Great hint thanks

**Comment by samcake** [comment]
This last commit should fix the problem on mac
WE now force the color mask to true when clearing the color buffers of a framebuffer
The StateCaceh does the rest.

THanks Austin!

**Comment by samcake** [comment]
I picked the selfieFaambuffer size to be HALF of the full window size. THis is arbitrary and was to save a bit of memory. IF we decide to use that framebuffer for something else we need to remember that !

**Comment by jherico** [comment]
we should not be using named framebuffers for the most part and not reusing framebuffers.  instead the framebuffercache should let you request and release framebuffers of a given size.

**Comment by hifi-gustavo** [comment]
Build console output for commit: e6f2504514313538689b721d7fa2b3e86c0df688 available at: https://gist.github.com/fee9f2c1417f6731faf7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5407.dmg

**Pull request merged**